python/lsst/daf/butler/formatters/file.py

```diff
@@ -104,35 +105,84 @@ def _assembleDataset(self, data: Any, component: Optional[str] = None) -> Any:
-                        f" converting type {get_full_type_name(fileDescriptor.storageClass.pytype)}"
-                        f" to {get_full_type_name(readStorageClass.pytype)}."
-                    )
-            else:
-                # Concrete composite written as a single file (we hope)
-                try:
-                    data = fileDescriptor.storageClass.delegate().getComponent(data, component)
-                except AttributeError:
-                    # Defer the complaint
-                    data = None
-
-        # Coerce to the requested type (not necessarily the type that was
-        # written)
-        data = self._coerceType(data, fileDescriptor.storageClass, readStorageClass)
+        writeStorageClass = fileDescriptor.storageClass
+
+        if component is not None:
+            # Requesting a component implies that we need to first ensure
+            # that the composite is the correct python type. Lie to the
+            # coercion routine since the read StorageClass is not relevant
+            # if we want the original.
+            data = self._coerceType(data, writeStorageClass, writeStorageClass)
+
+            # Concrete composite written as a single file (we hope)
+            # so try to get the component.
+            try:
+                data = fileDescriptor.storageClass.delegate().getComponent(data, component)
+            except AttributeError:
+                # Defer the complaint
+                data = None
+
+            # Update the write storage class to match that of the component.
+            # It should be safe to use the component storage class directly
+            # since that should match what was returned from getComponent
+            # (else we could create a temporary storage class guaranteed to
+            # match the python type we have).
+            writeStorageClass = writeStorageClass.allComponents()[component]
+
+        # Coerce to the requested type.
+        data = self._coerceType(data, writeStorageClass, readStorageClass)
 
         return data
 
+    def _coerceBuiltinType(self, inMemoryDataset: Any, writeStorageClass: StorageClass) -> Any:
+        """Coerce the supplied inMemoryDataset to the written python type if it
+        is currently a built-in type.
+
+        Parameters
+        ----------
+        inMemoryDataset : `object`
+            Object to coerce to expected type.
+        writeStorageClass : `StorageClass`
+            Storage class used to serialize this data.
+
+        Returns
+        -------
+        inMemoryDataset : `object`
+            Object of expected type ``writeStorageClass.pytype``.
+        """
+        # Only do anything if this is a built-in type and we did not originally
+        # store it as a builtin type.
+        if (
+            inMemoryDataset is not None
+            and not hasattr(builtins, writeStorageClass.pytype.__name__)
+            and hasattr(builtins, type(inMemoryDataset).__name__)
```

Comment:
JSON can return a list as well as a dict (or even a float or int on its own). The YamlFormatter can return a normal python type if the converter was not registered (and there is code in the YamlFormatter to try to convert the type to a dict).

It's possible I could change that third part of the `if` statement to be a limited set of `list|dict`. The logic is really "if we didn't start with a simple type but ended up reading one, try to convert it to the type that we started with"

Reply:
I've expanded the documentation for the method and cleaned up the logic further. The code now runs if:

* The written type does not match the type read from the file.
* The type read from the file looks like a built-in.

In theory that second check could be changed to be an explicit check for `list|dict|int|float|str` (types you can read from a JSON or YAML file without cleverness) but I'm not sure that really gives us anything.

Reply:
Could that be changed to:

- the written type does not match the type read from the file;
- `component=None`?

Or do we need the builtin checks to deal with storage class conversions that also have component extraction in play?

Reply:
Checking that the types don't match is not sufficient because then we can end up being very confused by pre-empting storage class conversions that happen immediately after these checks. These checks are here specifically for the case where you get a generic python type but want a more specialized type. Storage class conversions have to run after you've gone to the more specialized type.

component handling is handled above this method and if a component is in play I update the write storage class to match that of the component.